server/api/impl/store/manifest.go

```diff
@@ -0,0 +1,304 @@
+		}
+		if err != nil {
+			return errors.Wrapf(err, "failed to load global manifest for %s", string(appID))
+		}
+
+		var m *apps.Manifest
+		m, err = DecodeManifest(data)
+		if err != nil {
+			return err
+		}
+		if m.AppID != appID {
+			return errors.Errorf("mismatched app ids while getting manifest %s != %s", m.AppID, appID)
+		}
+		global[appID] = m
+	}
+
+	s.mutex.Lock()
+	s.global = global
+	s.mutex.Unlock()
+
+	return nil
+}
+
+func DecodeManifest(data []byte) (*apps.Manifest, error) {
+	var m apps.Manifest
+	err := json.Unmarshal(data, &m)
+	if err != nil {
+		return nil, err
+	}
+	err = validateManifest(&m)
+	if err != nil {
+		return nil, err
+	}
+	return &m, nil
+}
+
+func (s *manifestStore) Configure(conf api.Config) error {
+	updatedLocal := map[apps.AppID]*apps.Manifest{}
+
+	for id, key := range conf.LocalManifests {
+		var m *apps.Manifest
+		err := s.mm.KV.Get(api.PrefixLocalManifest+key, &m)
+		if err != nil {
+			s.mm.Log.Error(
+				fmt.Sprintf("failed to load local manifest for %s: %s", id, err.Error()))
+		}
+		if m == nil {
+			s.mm.Log.Error(
+				fmt.Sprintf("failed to load local manifest for %s: not found", id))
+		}
+
+		updatedLocal[apps.AppID(id)] = m
+	}
+
+	s.mutex.Lock()
+	s.local = updatedLocal
+	s.mutex.Unlock()
+
+	return nil
+}
+
+func (s *manifestStore) Get(appID apps.AppID) (*apps.Manifest, error) {
+	s.mutex.RLock()
+	builtin := s.builtin
+	local := s.local
+	global := s.global
+	s.mutex.RUnlock()
+
+	m, ok := builtin[appID]
+	if ok {
+		return m, nil
+	}
+	m, ok = local[appID]
+	if ok {
+		return m, nil
+	}
+	m, ok = global[appID]
+	if ok {
+		return m, nil
+	}
+	return nil, utils.ErrNotFound
+}
+
+func (s *manifestStore) AsMap() map[apps.AppID]*apps.Manifest {
+	s.mutex.RLock()
+	builtin := s.builtin
+	local := s.local
+	global := s.global
+	s.mutex.RUnlock()
+
+	out := map[apps.AppID]*apps.Manifest{}
+	for id, m := range global {
+		out[id] = m
+	}
+	for id, m := range local {
+		out[id] = m
+	}
+	for id, m := range builtin {
+		out[id] = m
+	}
+	return out
+}
+
+func (s *manifestStore) StoreLocal(m *apps.Manifest) error {
+	conf := s.conf.GetConfig()
+	prevSHA := conf.LocalManifests[string(m.AppID)]
+
+	data, err := json.Marshal(m)
+	if err != nil {
+		return err
+	}
+	sha := fmt.Sprintf("%x", sha1.Sum(data)) // nolint:gosec
+	if sha == prevSHA {
+		return nil
+	}
+
+	_, err = s.mm.KV.Set(api.PrefixLocalManifest+sha, m)
+	if err != nil {
+		return err
+	}
+
+	s.mutex.RLock()
+	local := s.local
+	s.mutex.RUnlock()
+	updatedLocal := map[apps.AppID]*apps.Manifest{}
+	for k, v := range local {
+		if k != m.AppID {
+			updatedLocal[k] = v
+		}
+	}
+	updatedLocal[m.AppID] = m
+	s.mutex.Lock()
+	s.local = updatedLocal
+	s.mutex.Unlock()
+
+	updated := map[string]string{}
+	for k, v := range conf.LocalManifests {
+		updated[k] = v
+	}
+	updated[string(m.AppID)] = sha
+	sc := *conf.StoredConfig
+	sc.LocalManifests = updated
+	err = s.conf.StoreConfig(&sc)
```

Comment:
why do we need to store the manifest in both the config and the KV store?

Reply:
We store the map of `ID`:`sha` in the config (to keep it small), and store the payload in the KV as `man_$(SHA)`.

Reply:
I'll add a comment at the top of the app.go and manifest.go files